main.py

```diff
@@ -236,104 +263,148 @@ def main(cl_arguments):
             # to be set to True. For mnli-diagnostic this flag will be False because it is part of GLUE and
             # "is_probing_task is global flag specific to a run, not to a task.
             if task.name == 'mnli-diagnostic':
                 continue
-            pred_module = getattr(model, "%s_mdl" % task.name)
-            to_train = elmo_scalars + [(n, p)
-                                       for n, p in pred_module.named_parameters() if p.requires_grad]
+
+            if args.transfer_paradigm == "finetune":
+                # Train both the task specific models as well as sentence encoder.
+                to_train = [(n, p) for n, p in model.named_parameters() if p.requires_grad]
+            else: # args.transfer_paradigm == "frozen":
+                # Only train task-specific module.
+                pred_module = getattr(model, "%s_mdl" % task.name)
+                to_train = [(n, p) for n, p in pred_module.named_parameters() if p.requires_grad]
+                to_train += elmo_scalars
+
+
             # Look for <task_name>_<param_name>, then eval_<param_name>
             params = build_trainer_params(args, task_names=[task.name, 'eval'])
             trainer, _, opt_params, schd_params = build_trainer(params, model,
                                                                 args.run_dir,
                                                                 task.val_metric_decreases)
-            best_epoch = trainer.train([task], task.val_metric,
-                                       args.batch_size, 1,
-                                       args.weighting_method, args.scaling_method,
-                                       to_train, opt_params, schd_params,
-                                       args.shared_optimizer, load_model=False, phase="eval")
+            best_epoch = trainer.train(tasks=[task], stop_metric=task.val_metric,
+                                       batch_size=args.batch_size, n_batches_per_pass=1,
+                                       weighting_method=args.weighting_method,
+                                       scaling_method=args.scaling_method,
+                                       train_params=to_train,
+                                       optimizer_params=opt_params,
+                                       scheduler_params=schd_params,
+                                       shared_optimizer=args.shared_optimizer,
+                                       load_model=False, phase="eval")
 
             # Now that we've trained a model, revert to the normal checkpoint logic for this task.
             task_names_to_avoid_loading.remove(task.name)
 
             # The best checkpoint will accumulate the best parameters for each task.
             # This logic looks strange. We think it works.
             best_epoch = best_epoch[task.name]
+            # TODO(Yada): Move logic for checkpointing finetuned vs frozen pretrained tasks
```

Comment:
@pruksmhc ?

Reply:
Move to trainer.py or its own file (see comment above w/ Sam)